whisper.cpp

```diff
@@ -3893,25 +3893,29 @@ static void whisper_suppress_invalid_grammar(
     }
 
     const auto rejects = whisper_grammar_reject_candidates(grammar.rules, grammar.stacks, candidates_grammar);
+
     for (const auto & reject : rejects) {
-        if (logits[reject.id] > 0) {
-            logits[reject.id] /= params.grammar_penalty;
-        } else {
-            logits[reject.id] *= params.grammar_penalty;
-        }
+        logprobs[reject.id] -= params.grammar_penalty;
```

Comment:
I think this makes more sense, though I'm still experimenting.
At least for the use case where we want to strongly restrict the output, this works well with large penalty (>100.0)

Reply:
Yeah, this seems to work well enough - I've been testing this code with 10 - 15 on tiny. I was hoping for a sweet spot that would guide the decoding towards valid strings while leaving obviously non-matching strings unchanged. But so far, I've found that the point that gives satisfying grammar matching still seems too eager to match unrelated words to the grammar.

Reply:
I suppose both variants to fail in some way for different use cases. This penalty is essentially a free-parameter which it's not obvious how to set. A more general way would be to let the user provide a penalty callback function and let them implement the penalization, but I guess for now lets keep it simple

> But so far, I've found that the point that gives satisfying grammar matching still seems too eager to match unrelated words to the grammar.

Does the original scaling work better in this scenario?

Reply:
> Does the original scaling work better in this scenario?

Anecdotally, no, not really.